src/Components/Components/test/Routing/RouterTest.cs

```diff
@@ -2,103 +2,245 @@
+
+            // Assert
+            Assert.True(called);
+            Assert.Single(_renderer.HandledExceptions);
+            var unhandledException = _renderer.HandledExceptions[0];
+            Assert.Equal("This is an uncaught exception.", unhandledException.Message);
+        }
+
+        [Fact]
+        public void CanceledFailedOnNavigateAsyncDoesNothing()
+        {
+            // Arrange
+            async Task OnNavigateAsync(NavigationContext args)
+            {
+                if (args.Path.EndsWith("jan"))
+                {
+                    await Task.Delay(Timeout.Infinite);
+                    throw new Exception("This is an uncaught exception.");
+                }
+            }
+            var refreshCalled = false;
+            _renderer.OnUpdateDisplay = (renderBatch) =>
+            {
+                if (!refreshCalled)
+                {
+                    Assert.True(true);
+                    return;
+                }
+                Assert.True(false, "OnUpdateDisplay called more than once.");
+            };
+            _router.OnNavigateAsync = new EventCallbackFactory().Create<NavigationContext>(_router, OnNavigateAsync);
+
+            // Act
+            _ = _renderer.Dispatcher.InvokeAsync(() => _router.RunOnNavigateWithRefreshAsync("http://example.com/jan", false));
+            _ = _renderer.Dispatcher.InvokeAsync(() => _router.RunOnNavigateWithRefreshAsync("http://example.com/feb", false));
+
+            // Assert that we render the second route component and don't throw an exception
+            Assert.Empty(_renderer.HandledExceptions);
+        }
+
+        [Fact]
+        public async Task CanHandleSingleCancelledOnNavigateAsync()
+        {
+            // Arrange
+            async Task OnNavigateAsync(NavigationContext args)
+            {
+                var tcs = new TaskCompletionSource<int>();
+                tcs.TrySetCanceled();
+                await tcs.Task;
+            }
+            _renderer.OnUpdateDisplay = (renderBatch) => Assert.True(false, "OnUpdateDisplay called more than once.");
+            _router.OnNavigateAsync = new EventCallbackFactory().Create<NavigationContext>(_router, OnNavigateAsync);
+
+            // Act
+            var janTask = _router.RunOnNavigateWithRefreshAsync("http://example.com/jan", false);
+
+            var janTaskException = await Assert.ThrowsAsync<InvalidOperationException>(() => janTask);
+
+            // Assert
+            Assert.Equal("OnNavigateAsync callback cannot be canceled.", janTaskException.Message);
+        }
+
+        [Fact]
+        public async Task AlreadyCanceledOnNavigateAsyncDoesNothing()
+        {
+            // Arrange
+            async Task OnNavigateAsync(NavigationContext args)
+            {
+                if (args.Path.EndsWith("jan"))
+                {
+                    var tcs = new TaskCompletionSource();
+                    await Task.Delay(4000);
```

Comment:
There might be an opportunity to avoid blocking the test run for 4 seconds if this was `await someOtherTaskCompletionSource.Task` where `someOtherTaskCompletionSource` is defined at the top level in the `AlreadyCanceledOnNavigateAsyncDoesNothing`, and whose completion is triggered right after line 164 below.